transport/handler_server.go

```diff
@@ -227,6 +231,9 @@ func (ht *serverHandlerTransport) WriteStatus(s *Stream, st *status.Status) erro
 				}
 			}
 		}
+		if ht.stats != nil {
```

Comment:
A proposal to resolve the possible race you mentioned: move this `HandleRPC` call out from this `func` and call it after `ht.do` returns:
```go
err := ht.do(func() {})
if err == nil {
  if ht.stats != nil {
    ht.stats.HandleRPC(s.Context(), &stats.OutTrailer{})
  }
  // Other stuff
}
```

This should work because the `func` is scheduled by `ht.do` doesn't return in the middle. So we can assume it would eventually happen, it would be OK to handle `OutTrailer` before it actually happens.